homeassistant/components/netatmo/data_handler.py

```diff
@@ -0,0 +1,170 @@
+        )
+
+        self.listeners.append(
+            self.hass.bus.async_listen("netatmo_event", self.handle_event)
+        )
+
+    async def async_update(self, event_time):
+        """
+        Update device.
+
+        We do up to BATCH_SIZE calls in one update in order
+        to minimize the calls on the api service.
+        """
+        for data_class in islice(self._queue, 0, BATCH_SIZE):
+            if data_class["next_scan"] > time():
+                continue
+            self._data_classes[data_class["name"]]["next_scan"] = (
+                time() + data_class["interval"]
+            )
+
+            await self.async_fetch_data(
+                data_class["class"], data_class["name"], **data_class["kwargs"]
+            )
+
+        self._queue.rotate(BATCH_SIZE)
+
+    async def async_cleanup(self):
+        """Clean up the Netatmo data handler."""
+        for listener in self.listeners:
+            listener()
+
+    async def handle_event(self, event):
+        """Handle webhook events."""
+        if event.data["data"]["push_type"] == "webhook_activation":
+            _LOGGER.info("%s webhook successfully registered", MANUFACTURER)
+            self._webhook = True
+
+    async def async_fetch_data(self, data_class, data_class_entry, **kwargs):
+        """Fetch data and notify."""
+        try:
+            self.data[data_class_entry] = await self.hass.async_add_executor_job(
+                partial(data_class, **kwargs), self._auth,
+            )
+            async_dispatcher_send(self.hass, f"netatmo-update-{data_class_entry}")
+        except (pyatmo.NoDevice, pyatmo.ApiError) as err:
+            _LOGGER.debug(err)
+
+    async def register_data_class(self, data_class_name, **kwargs):
+        """Register data class."""
+        if "home_id" in kwargs:
+            data_class_entry = f"{data_class_name}-{kwargs['home_id']}"
+        elif "area_name" in kwargs:
+            data_class_entry = f"{data_class_name}-{kwargs.pop('area_name')}"
+        else:
+            data_class_entry = data_class_name
+
+        async with self.lock:
+            if data_class_entry not in self._data_classes:
+                self._data_classes[data_class_entry] = {
+                    "class": DATA_CLASSES[data_class_name],
+                    "name": data_class_entry,
+                    "interval": DEFAULT_INTERVALS[data_class_name],
+                    "next_scan": time() + DEFAULT_INTERVALS[data_class_name],
+                    "kwargs": kwargs,
+                    "registered": 1,
+                }
+
+                await self.async_fetch_data(
+                    DATA_CLASSES[data_class_name], data_class_entry, **kwargs
+                )
+
+                self._queue.append(self._data_classes[data_class_entry])
+                _LOGGER.debug("Data class %s added", data_class_entry)
+
+            else:
+                self._data_classes[data_class_entry].update(
```

Comment:
Maybe instead of the ref counting we could store a callback and return a handle that unsubscribed from the data class.

This would avoid having to watch for all the event as we could process the callbacks when there was an updated instead.

That is a big change so probably for another PR though.

Reply:
Since there is still a bit time until 0.114 I'll give it a try if this is (one of) the last thing(s) to be addressed.

Reply:
Looks like this is really the last item of substance.
`homeassistant/helpers/update_coordinator.py` and `august/subscriber.py` might be inspirational. (update_coordinator handles unsubscribing better)

Reply:
I'll take a look at august. `update_coordinator` unfortunately was not an option due to the usage of webhook responses.